Stripe/STPPaymentMethodsInternalViewController.m

```diff
@@ -9,125 +9,315 @@
-    if (section == STPPaymentMethodCardListSection) {
+    if (section == PaymentMethodSectionCardList) {
         return self.paymentMethods.count;
-    } else if (section == STPPaymentMethodAddCardSection) {
+    }
+
+    if (section == PaymentMethodSectionAddCard) {
         return 1;
     }
+
     return 0;
 }
 
 - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
-    STPPaymentMethodTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:STPPaymentMethodCellReuseIdentifier forIndexPath:indexPath];
-    if (indexPath.section == STPPaymentMethodCardListSection) {
+    STPPaymentMethodTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:PaymentMethodCellReuseIdentifier forIndexPath:indexPath];
+
+    if (indexPath.section == PaymentMethodSectionCardList) {
         id<STPPaymentMethod> paymentMethod = [self.paymentMethods stp_boundSafeObjectAtIndex:indexPath.row];
-        [cell configureWithPaymentMethod:paymentMethod
-                                selected:[paymentMethod isEqual:self.selectedPaymentMethod]
-                                   theme:self.theme];
-    } else {
+        BOOL selected = [paymentMethod isEqual:self.selectedPaymentMethod];
+
+        [cell configureWithPaymentMethod:paymentMethod theme:self.theme selected:selected];
+    }
+    else {
         [cell configureForNewCardRowWithTheme:self.theme];
     }
+
     return cell;
 }
 
+- (BOOL)tableView:(__unused UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
+    if (indexPath.section == PaymentMethodSectionCardList) {
+        id<STPPaymentMethod> paymentMethod = [self.paymentMethods stp_boundSafeObjectAtIndex:indexPath.row];
+
+        if ([self isPaymentMethodDetachable:paymentMethod]) {
+            return YES;
+        }
+    }
+
+    return NO;
+}
+
+- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
+    if (indexPath.section == PaymentMethodSectionCardList) {
+        if (editingStyle != UITableViewCellEditingStyleDelete) {
+            // Showed the user a non-delete option when we shouldn't have
+            [tableView reloadData];
+            return;
+        }
+
+        if (!(indexPath.row < (NSInteger)self.paymentMethods.count)) {
+            // Data source and table view out of sync for some reason
+            [tableView reloadData];
+            return;
+        }
+
+        id<STPPaymentMethod> paymentMethodToDelete = [self.paymentMethods stp_boundSafeObjectAtIndex:indexPath.row];
+
+        if (![self isPaymentMethodDetachable:paymentMethodToDelete]) {
+            // Showed the user a delete option for a payment method when we shouldn't have
+            [tableView reloadData];
+            return;
+        }
+
+        if (![paymentMethodToDelete conformsToProtocol:@protocol(STPSourceProtocol)]) {
```

Comment:
yeah it is but i think it's not too much extra work to have an additional protocol check closer to the casting just in case the implementation of `isPaymentMethodDetachable` changes dramatically.